### PR TITLE
Ensure journal is empty

### PR DESCRIPTION
**Pull request by sergey-kintsel**
Refactor `Chewy::Journal`
Add multiple retries to apply the journal

### CODE REVIEW

**Comment by DNNX** [comment]
The code looks ok. I sent you a few small remarks in DM. Let's :shipit: after green build.

File: lib/chewy/fields/base.rb [hunk 1]
```diff
@@ -51,11 +51,13 @@ def compose(object, *parent_objects)
           object.send(name)
         end
 
-        result = if result.respond_to?(:to_ary)
-          result.to_ary.map { |item| compose_children(item, *objects) }
-        else
-          compose_children(result, *objects)
-        end if children.present? && !multi_field?
+        if children.present? && !multi_field?
```
Comment (pr0d1r2) [inline lib/chewy/fields/base.rb]: Please consider splitting whole `compose` method as my eyes are bleeding:

```ruby
def compose(object, *parent_objects)
  { name => compose_result(object, parent_objects) }
end

def compose_result(object, *parent_objects)
  objects = ([object] + parent_objects.flatten).uniq
  result = base_result(object, objects)
  return compose_result_with_children(result, objects) if children.present? && !multi_field?
  result
end

def base_result(object, objects)
  if value && value.is_a?(Proc)
    result_from_proc(object, objects)
  elsif object.is_a?(Hash)
    result_from_hash(object)
  else
    object.send(name)
  end
end

def result_from_proc(object, objects)
  if value.arity.zero?
    object.instance_exec(&value)
  elsif value.arity < 0
    value.call(*object)
  else
    value.call(*objects.first(value.arity))
  end
end

def result_from_hash(object)
  if object.key?(name)
    object[name]
  else
    object[name.to_s]
  end
end

def compose_result_with_children(result, objects)
  if result.respond_to?(:to_ary)
    result.to_ary.map { |item| compose_children(item, *objects) }
  else
    compose_children(result, *objects)
  end
end
```

File: lib/chewy/index.rb [hunk 2]
```diff
@@ -90,14 +90,13 @@ def self.define_type(target, options = {}, &block)
...
-          def self.#{type_class.type_name}
-            ActiveSupport::Deprecation.warn("`#{self}.#{type_class.type_name}` accessor is deprecated and will be removed soon. Use `#{type_class}` directly instead.")
-            type_hash['#{type_class.type_name}']
-          end
-        METHOD
-      end
+      return if respond_to?(type_class.type_name)
+      class_eval <<-METHOD, __FILE__, __LINE__ + 1
+        def self.#{type_class.type_name}
+          ActiveSupport::Deprecation.warn("`#{self}.#{type_class.type_name}` accessor is deprecated and will be removed soon. Use `#{type_class}` directly instead.")
```
Comment (pr0d1r2) [inline lib/chewy/index.rb]: Please break this long line by using heredoc:

```ruby
ActiveSupport::Deprecation.warn(<<WARNING.sqish)
  `#{self}.#{type_class.type_name}` accessor is deprecated and will be removed soon.
  Use `#{type_class}` directly instead.
WARNING

```

File: lib/chewy/journal/clean.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+module Chewy
+  class Journal
+    module Clean
+      DELETE_BATCH_SIZE = 10_000
+
+      def until(time)
+        query = Query.new(time, :lte, nil, false).to_h
+        search_query = query.merge(fields: ['_id'], size: DELETE_BATCH_SIZE)
+        index_name = Journal.index_name
```
Comment (pr0d1r2) [inline lib/chewy/journal/clean.rb]: We can use delegation:

```ruby
included do
  delegate :index_name, to: Journal
end
```

File: spec/chewy/journal/apply_spec.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            default_import_options journal: true
+          end
+        end
+        stub_index('country') do
+          define_type Country do
+            default_import_options journal: true
+          end
+        end
+
+        Chewy.massacre
```
Comment (pr0d1r2) [inline spec/chewy/journal/apply_spec.rb]: 😅

File: lib/chewy/index.rb (hunk 2 above)
Comment (sergey-kintsel) [inline lib/chewy/index.rb]: it is not connected to my changes) this was changed to make Rubocop happy

File: lib/chewy/fields/base.rb (hunk 1 above)
Comment (sergey-kintsel) [inline lib/chewy/fields/base.rb]: it is not connected to my changes) this was changed to make Rubocop happy

File: lib/chewy/journal/clean.rb (hunk 3 above)
Comment (sergey-kintsel) [inline lib/chewy/journal/clean.rb]: actually, I think that explicit calls are easier) longer, but easier)

Comment (pr0d1r2) [inline lib/chewy/journal/clean.rb]: OK

File: lib/chewy/fields/base.rb (hunk 1 above)
Comment (pr0d1r2) [inline lib/chewy/fields/base.rb]: Can you please prepare follow-up PR after merging this one?

File: lib/chewy/index.rb (hunk 2 above)
Comment (pr0d1r2) [inline lib/chewy/index.rb]: Can you please prepare follow-up PR after merging this one?

**Pull request merged**